packages/giselle/src/engine/generations/generate-text.ts

```diff
@@ -171,16 +170,28 @@ export function generateText(args: {
 			if (
 				operationNode.content.llm.provider === "google" &&
 				operationNode.content.llm.configurations.searchGrounding &&
-				hasCapability(languageModel, Capability.OptionalSearchGrounding)
+				hasCapability(languageModel, Capability.OptionalSearchGrounding) &&
+				operationNode.content.contextSource === "google_search"
 			) {
 				preparedToolSet = {
 					...preparedToolSet,
 					toolSet: {
 						...preparedToolSet.toolSet,
-						google_search: googleTools.googleSearch({}),
+						google_search: vertex.tools.googleSearch({}),
 					},
 				};
 			}
+			if (
+				operationNode.content.llm.provider === "google" &&
+				hasCapability(languageModel, Capability.UrlContext) &&
+				operationNode.content.contextSource === "url_context"
+			) {
+				preparedToolSet = addUrlContextTool({
+					preparedToolSet,
+					urls: urlContextUrls,
+					tool: vertex.tools.urlContext({}),
+				});
```

Comment:
As with google_search, I think it would be enough to just add it to the tool if url_context is on in the model configuration, but I don't understand the purpose of inserting the addUrlContextTool function.